### ISSUE TITLE
Building Stylelint for client-side

### ISSUE THREAD

**Issue by oskarkrawczyk** [stylelint/stylelint#3935]
I'm currently trying to replace CSSLint in @jsfiddle with Stylelint. Got it working thanks to a proof-of-concept made here https://github.com/m-allanson/stylelint-browser-demo

Unfortunately turns out this works fine but with a very old version of Stylelint, specifically the 5.x branch. Once updated deps to the newest Stylelint, the whole client-side bundle breaks apart with a slightly cryptic `Error: Cannot find module "."` error.

I was wondering if anyone around here can help out figure out if it's even possible to run the newest Stylelint on the client-side (this is a requirement for JSFiddle as our env only supports linting that way).

---

A bit more on what my config looks like:

`webpack.config`:

```js
import webpack from "webpack"
import path from "path"

export default {
  entry: "./site/index.js",
  output: {
    path: path.resolve(__dirname, "site"),
    publicPath: "/",
    filename: "bundle.js",
  },
  resolve: {
    root: path.resolve(__dirname),
    alias: {
      "resolve-from": "empty-module",
      cosmiconfig: "empty-module",
      doiuse: "empty-module",
      globby: "empty-module",
      globjoin: "empty-module",
      multimatch: "empty-module",
      path: "empty-module",
    },
    modulesDirectories: [
      "node_modules",
      "site",
    ]
  },
  module: {
    loaders: [
      {
        test: /\.js$/,
        loader: "babel-loader",
        exclude: /node_modules/,
      },
      {
        test: /\.json$/,
        loader: "json-loader",
      },
    ],
  },
  plugins: [
    new webpack.optimize.DedupePlugin(),
    new webpack.ContextReplacementPlugin(/stylelint/, /NEVER_MATCH^/),
  ],
  node: {
    fs: "empty",
    module: "empty",
  },
}
```

`package.json`:

```json
{
  "name": "stylelint-browser-demo",
  "version": "0.0.1",
  "description": "A proof-of-concept to demo stylelint running in a browser.",
  "main": "site/bundle.js",
  "scripts": {
    "start": "npm run build && http-server ./site",
    "start-dev": "webpack-dev-server --content-base site/ --hot --inline",
    "build": "webpack --progress --colors && stat -f%z site/bundle.js",
    "lint": "eslint . --ignore-path .gitignore",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "m-allanson",
  "license": "MIT",
  "dependencies": {
    "babel-cli": "^6.1.18",
    "babel-loader": "^6.2.4",
    "babel-preset-es2015": "^6.6.0",
    "http-server": "^0.9.0",
    "imports-loader": "^0.6.5",
    "json-loader": "^0.5.4",
    "stylelint": "^9.10.1",
    "stylelint-config-standard": "^18.2.0",
    "webpack": "^1.12.14"
  },
  "devDependencies": {
    "eslint": "^2.4.0",
    "eslint-config-stylelint": "^1.0.0",
    "webpack-dev-server": "^1.14.1"
  },
  "engines": {
    "node": "5.9.1",
    "npm": "3.7.3"
  },
  "eslintConfig": {
    "extends": "stylelint",
    "env": {
      "browser": true
    },
    "rules": {
      "arrow-spacing": 2,
      "no-var": 2,
      "object-shorthand": 2,
      "prefer-const": 2,
      "template-curly-spacing": 2
    }
  },
  "babel": {
    "presets": [
      "babel-preset-es2015"
    ]
  }
}
```

The build `bundle.js` is attached - [bundle.js.zip](https://github.com/stylelint/stylelint/files/2821391/bundle.js.zip)

Lint method is exposed as `Stylelint.verify()`

@ai Any chance you could help out? (we had a brief chat about this on Twitter).

**Comment by jeddy3** [comment]
> Unfortunately turns out this works fine but with a very old version of Stylelint, specifically the 5.x branch

We ran into our own difficulties trying to get stylelint to run client-side. We ended up going server-side for [our demo on the website](https://github.com/stylelint/stylelint-demo). However, this was a couple of years ago and the tooling might have moved on enough to take another crack at it.

> Seems like we have too dymanic require somewhere.

The heaviest parts of stylelint are the:
- parsers
- rules

I believe we use dynamic requires for these.

They were added to improve the bad performance outlined in https://github.com/stylelint/stylelint/issues/2454. However, there's probably a better way. One that improves performance _and_ caters for the client side use case.

---

I've made a couple of attempts recently to use [`ncc`](https://github.com/zeit/ncc) but the load-time of stylelint seemed to worsen so I didn't pursue it.

A good way forward might be to adopt [prettier's approach to building](https://github.com/prettier/prettier/tree/master/scripts/build). They are dealing with similar problems to us:

- the need for a standalone (client-side) build
- bundling a number of (heavy) parsers

And they have far more contributors dedicating their time to the problem.

Or it might be that unpicking some of our dynamic requires to support your webpack config will require less effort and time.

Has anyone else got any ideas?

**Comment by ai** [comment]
I think we can have internal API for client-side and auto-load API for CLI:

```js
// Client-side

let core = require('stylelint/…')
let scss = require('postcss-scss')

core.check(source, { parser: scss })
```

```js
// CLI
let autoloader = require('stylelint/…')

autoloader.autodetect(source, filename) // will require core and scss dynamically inside
```

**Comment by oskarkrawczyk** [comment]
> I'm going to label this issue up as an enhancement and help wanted

Sounds good! I'll push this issue though JSFiddle channel, perhaps someone will have the resources to look into this. If not, do you mind if I post it as a bounty? I want to be 100% sure you're ok with that.

**Comment by jeddy3** [comment]
> If not, do you mind if I post it as a bounty?

Sounds good to me!

**Comment by ai** [comment]
I will promote this issue in PostCSS twitter too.

**Comment by konpikwastaken** [comment]
Couldn't sleep, wanted to see if it could be hacked together to run similarly to the POC done earlier. Hopefully this might help someone in the future.

**tl;dr; it's possible in the current state**

https://github.com/konpikwastaken/stylelint-browser-poc

**Comment by evilebottnawi** [comment]
@konpikwastaken great job :+1:

**Comment by m-allanson** [comment]
It's great to see work happening around this!

@konpikwastaken's comment on their [linked repo](https://github.com/konpikwastaken/stylelint-browser-poc) is possibly the key to making this work well:

> The engine should be refactored & decoupled from anything that depends on an actual file system (e.g. reading configs, etc). Majority of the time was spent working around assumptions that the execution environment has IO access (e.g. path, fs).

**Comment by jeddy3** [comment]
> The engine should be refactored & decoupled from anything that depends on an actual file system (e.g. reading configs, etc).

Sounds great! 

I think we can close this issue when someone finds time to do that refactoring.

**Comment by jeddy3** [comment]
> The engine should be refactored & decoupled from anything that depends on an actual file system (e.g. reading configs, etc). Majority of the time was spent working around assumptions that the execution environment has IO access (e.g. path, fs).

@m-allanson and I are going to look into this.

It'd be great if we ended up with something similar to how [Prettier does it](https://prettier.io/docs/en/browser.html):

```css
<script src="https://unpkg.com/stylelint@14.0.0/universal.js"></script>
<script src="https://unpkg.com/stylelint@14.0.0/syntax-scss.js"></script>
<script>
  stylelint.lint("a#{var} { color: red; }", {
    config: { "rules": { "color-named"} },
    syntax: "scss",
  });
</script>
```

The issue is related to https://github.com/stylelint/stylelint/issues/2454, where we need to resolve some performance issues.

**Comment by m-allanson** [comment]
> @m-allanson and I are going to look into this.

In fact we have already spent some time working on this 😄. You can see a live demo of progress here: https://upbeat-yonath-79931e.netlify.app

I think some refactoring of stylelint's internals would make this quite achievable, and would allow us to avoid previous approaches of stubbing out modules that access the filesystem.

[Check out this repo to see how the demo works](https://github.com/m-allanson/sl-basic-runner). It's a relatively small amount of code.

**Comment by m-allanson** [comment]
Following @jeddy3's changes in #4729 I've got a WIP branch here: https://github.com/stylelint/stylelint/compare/browser-bundle. It's still pretty rough but demonstrates that this should work. 

Syntax parsers are bundled separately and can be loaded on demand. 

The stylelint bundle comes out at 250KB (minified + gzipped), with the syntax parsers ranging between 30KB and 260KB. 

| bundle    | min + gzip | min only |
|-----------|------------|----------|
| stylelint | 251K       | 1.3MB    |
| CSS in JS | 260K       | 1.0MB    |
| HTML      | 120K       | 433K     |
| Less      | 31K        | 106K     |
| Markdown  | 145K       | 509K     |
| Sass      | 71K        | 287K     |
| Scss      | 32K        | 109K     |
| SugarSS   | 33K        | 113K     |

There's an updated demo at https://stylelint-browser-bundle.netlify.app/. Check the network tab when selecting a syntax to see parsers being loaded in as needed.

**Comment by m-allanson** [comment]
I've opened a draft PR that details the current progress of this feature: https://github.com/stylelint/stylelint/pull/4796

**Comment by agray5** [comment]
Any updates on this?

**Comment by adalinesimonian** [comment]
VS Code now [runs in the browser](https://code.visualstudio.com/updates/v1_62#_visual-studio-code-for-the-web-vscodedev-preview) as well as on desktop. In theory, if browser support ever does get implemented, it may then be possible to have the VS Code extension run in the browser as well since vscode-languageserver [has browser support](https://github.com/microsoft/vscode-languageserver-node#3160-protocol-600-json-rpc-700-client-and-700-server).

**Comment by jeddy3** [comment]
I saw that. Very exciting stuff.

I don't think we've far off. @m-allanson has mostly got us there. I think we'll need to:
- https://github.com/stylelint/stylelint/issues/5291
- Remove processors (which can only happen when there's a robust and maintained styled-components custom syntax)

**Comment by adalinesimonian** [comment]
> I don't think we've far off. @m-allanson has mostly got us there. I think we'll need to:
> 
>     * [Move to ESM #5291](https://github.com/stylelint/stylelint/issues/5291)

I'll need to investigate a few things downstream. I remember I was having issues using ESM in VS Code while testing. However, I don't really remember exactly what went wrong. I'll take a look into it, maybe after I take care of a few issues I'm working on [in v1.2](https://github.com/stylelint/vscode-stylelint/projects/3).

**Comment by ota-meshi** [comment]
I'm experimenting with getting stylelint to work on browser.
Demo: https://ota-meshi.github.io/stylelint4b/Playground.html

I now realize that there are a lot of tricks required to get stylelint working on the browser.
I would like to share those information.

- Stub

We need to replace some file-related external modules with modules that work on the browser.
https://github.com/ota-meshi/stylelint4b/blob/e210f823992126c610c144fb3e3da817167b2c73/packages/stylelint4b/build/index.js#L51

Also, stylelint's internal module needs to be replaced with modules that works on the browser.

https://github.com/ota-meshi/stylelint4b/blob/e210f823992126c610c144fb3e3da817167b2c73/packages/stylelint4b/build/index.js#L17

- Replace source code

Some source code needs to be replaced and rewritten for the bundle.

https://github.com/ota-meshi/stylelint4b/blob/e210f823992126c610c144fb3e3da817167b2c73/packages/stylelint4b/build/index.js#L35

- Configuration and plugin resolution

It hacks the require function to resolve plugins and shareable configs.

https://github.com/ota-meshi/stylelint4b/blob/e210f823992126c610c144fb3e3da817167b2c73/packages/stylelint4b/src/require-shim.js

Users can resolve plugins and shareable configurations by pre-registering aliases.

https://ota-meshi.github.io/stylelint4b/stylelint4b/#stylelint4b-alias
https://github.com/ota-meshi/stylelint4b/blob/e210f823992126c610c144fb3e3da817167b2c73/packages/stylelint4b/src/alias-module.js

Unless you use plugins or shareable configs, we probably don't need it.

**Comment by ybiquitous** [comment]
@ota-meshi Thanks for the share. That's an interesting trick! I now understand well what we need for running Stylelint on browsers.

**Comment by abose** [comment]
Hi all, I was trying to integrate stylelint too for [phcode.dev](https://phcode.dev), the browser version of [braclkerts.io](https://braclkerts.io) code editor.

Is it still not supported/how much work is left to port it to the browser?
Many thanks for considering my request.

**Comment by eight04** [comment]
Stylus is a usercsss browser extension, which uses stylelint as the CSS linter in the editor. Therefore we have a prebuilt bundle here:
https://github.com/openstyles/stylelint-bundle

**Comment by bartveneman** [comment]
I'd also like to chime in, because I'm attempting to run Stylelint in the browser for [Project Wallace's linter page](https://www.projectwallace.com/lint-css). The v18 branch already fixed some issues, like getting rid of some Node/Server-only modules! Thanks so much already for that! ❤️ 

## Use case summary

I want to run Stylelint in the browser, where I already have CSS available, either via scraping or direct input. This means I have no I/O and I don't need any globbing, but I do have the CSS already available in one or more strings/objects/`File[]`/whatever.
Another benefit here would be an option to run https://stylelint.io/demo/ without web containers!

## What I'd like

The best option for me would be to have this sort of public API that would help me run Stylelint client-side on projectwallace.com:

- Be able to compose my own lint pipeline: call `lintPostcssResult()`, `normalizeAllRuleSettings()`, `createPartialStylelintResult()`, `prepareReturnValue()` without reliance on node/server API's like `node:os` or `node:process`, _or_ with sensible fallbacks
- Be able to skip all read-from-disk and write-to-stdout things and only focus on the core: feed it a config, CSS and see a result come out
- Being able to call `createPlugin()` and `utils.{report,ruleMessages,validateOptions}` without that pulling in all of the server-only code

I know this is quite a big ask, but I think (with my _very_ limited knowledge!) that it could be done without harming the overall goals of Stylelint. For me it'd be just a cleaner separation of concerns, which should make testing a little easier as well.

_Note: I had an LLM help me figure out which would be the entry points for me to call to make this happen, so my apologies if I didn't list the correct ones. I did check them myself and it seems like these are things I'd need exposed to make my case work._

P.S.: thanks to @romainmenke for thinking along with me [on Bluesky](https://bsky.app/profile/romainmenke.bsky.social/post/3mvx6dviwnc2w) ❤️
